### ISSUE TITLE
Supervised learning pipeline

### ISSUE THREAD

**Issue by ttrine** [Rochester-NRT/AlphaGo#58]
I'm working away on this in the training branch. Should be done in the next few days.
Todo:
- [x] Symmetries. Randomly transform training samples to one of 8 possible reflection/rotation combinations, this mitigates positional overfitting.
- [x] Perspective invariance. All training should occur from the perspective of black.
- [x] Serialization. Save model each epoch.
- [ ] Parallel training. Training should take advantage of however many GPUs we throw at it.
- [ ] Testing. Needs better testing with assertions etc.

**Comment by AntreasAntoniou** [comment]
Hey man, Can you give me a brief of the training branch?

**Comment by ttrine** [comment]
Sure! Its main divergence from Develop is some refactoring, which you can get a feel for at #46 .
Apart from that, I'm building out the training pipeline in AlphaGo/training/supervised_policy_trainer.py. 
The basic workflow is:
1. Instantiate a `trainer = supervised_policy_trainer(**hyperparams)` object with training hyperparameters such as learning rate.
2. Run `trainer.train(trainset_folder,testset_folder,model_folder)` to train on samples from the train folder, validate on samples from the test folder, and, each epoch, save models to the model folder.

Once it's passing its basic test (tests/test_supervised_policy_trainer.py), I plan to work on the features in the above todo list.

Did I miss anything?

**Comment by AntreasAntoniou** [comment]
Ok. So where should I aim my efforts?

**Comment by ttrine** [comment]
Thanks for your enthusiasm! Until I merge the training branch with develop, I'm responsible for designing and implementing supervised policy network training. Once it's merged, there will definitely be hidden bugs and some nice ways we could extend/generalize the pipeline. So keep your eyes peeled! I should be done with the core supervised training pipeline by Monday.
In the meantime, do either of the following tickets grab your interest? 
- At #48, one of the most pressing needs for us is to formally specify the fast rollout policy. It seems LetterRip has been working on this, you could ask if he needs any help.
- Equally important before we can train, if you pop over to #51, our process for converting sgf files into GameState objects is incomplete.

**Comment by AntreasAntoniou** [comment]
@ttrine I was wondering what is available for me to work on for today? Finally on leave for a few days :)
